File: packages/toolgovern/src/classifier/filesystem-scope.ts
/**
 * TG02 -- Filesystem Scope Escalation
 *
 * Fires when a call attempts a write, delete, or permission change outside the caller's
 * declared filesystem scope (`scope.filesystem`, a list of allowed path prefixes), or targets a
 * small set of sensitive absolute system directories regardless of scope.
 */

import type { Rule, RuleMatch } from '../types.js';
import { containsPathTraversal, extractOperation, extractPath, isPathWithin } from './util.js';

const category = 'TG02' as const;

const WRITE_OPS = new Set(['write', 'create', 'append', 'put', 'save']);
const DELETE_OPS = new Set(['delete', 'remove', 'unlink', 'rm', 'rmdir']);
const CHMOD_OPS = new Set(['chmod', 'chown', 'setpermissions', 'set_permissions']);
const READ_OPS = new Set(['read', 'get', 'load', 'fetch', 'cat', 'open']);
const SENSITIVE_SYSTEM_PREFIXES = ['/etc', '/usr', '/bin', '/sbin', '/system', '/private/etc'];

function isWithinScope(path: string, filesystem: readonly string[]): boolean {
  if (filesystem.length === 0) return false;
  return filesystem.some((prefix) => isPathWithin(path, prefix));
}

function match(
  rule: Pick<Rule, 'id' | 'category'>,
  decision: RuleMatch['decision'],
  reason: string,
  matchedArgument: string,
): RuleMatch {
  return { ruleId: rule.id, category: rule.category, decision, reason, matchedArgument };
}

const writeOutsideScope: Rule = {
  id: 'TG02-write-outside-scope',
  category,
  description: 'A write/create targets a path outside the declared filesystem scope.',
  evaluate(ctx) {
    const path = extractPath(ctx.args);
    if (!path) return null;
    const op =
      extractOperation(ctx.args) ?? (ctx.tool.toLowerCase().includes('write') ? 'write' : '');
    if (!WRITE_OPS.has(op)) return null;
    if (isWithinScope(path, ctx.scope.filesystem)) return null;
    return match(
      this,
      'require-approval',
      `Write target "${path}" is outside the declared filesystem scope.`,
      path,
    );
  },
};

const deleteOutsideScope: Rule = {
  id: 'TG02-delete-outside-scope',
  category,
  description: 'A delete targets a path outside the declared filesystem scope.',
  evaluate(ctx) {
    const path = extractPath(ctx.args);
    if (!path) return null;
    const op =
      extractOperation(ctx.args) ?? (ctx.tool.toLowerCase().includes('delete') ? 'delete' : '');
    if (!DELETE_OPS.has(op)) return null;
    if (isWithinScope(path, ctx.scope.filesystem)) return null;
    return match(
      this,
      'deny',
      `Delete target "${path}" is outside the declared filesystem scope.`,
      path,
    );
  },
};

const chmodOutsideScope: Rule = {
  id: 'TG02-chmod-outside-scope',
  category,
  description: 'A permission change targets a path outside the declared filesystem scope.',
  evaluate(ctx) {
    const path = extractPath(ctx.args);
    if (!path) return null;
    const op = extractOperation(ctx.args);
    if (!op || !CHMOD_OPS.has(op)) return null;
    if (isWithinScope(path, ctx.scope.filesystem)) return null;
    return match(
      this,
      'deny',
      `Permission change on "${path}" is outside the declared filesystem scope.`,
      path,
    );
  },
};

const readOutsideScope: Rule = {
  id: 'TG02-read-outside-scope',
  category,
  description:
    "A read targets a path outside the caller's declared filesystem scope. Only fires when a " +
    'filesystem boundary was actually declared -- reads are not flagged for a caller with no ' +
    'declared filesystem scope at all (that absence is enforced elsewhere, e.g. TG05 zero-' +
    'capability denial), since nothing here would tell a read apart from any other unscoped call.',
  evaluate(ctx) {
    if (ctx.scope.filesystem.length === 0) return null;
    const path = extractPath(ctx.args);
    if (!path) return null;
    const op =
      extractOperation(ctx.args) ?? (ctx.tool.toLowerCase().includes('read') ? 'read' : '');
    if (!READ_OPS.has(op)) return null;
    if (isWithinScope(path, ctx.scope.filesystem)) return null;
    return match(
      this,
      'require-approval',
      `Read target "${path}" is outside the declared filesystem scope.`,
      path,
    );
  },
};

const pathTraversal: Rule = {
  id: 'TG02-path-traversal',
  category,
  description: 'A path uses ".." segments that could escape a scoped prefix.',
  evaluate(ctx) {
    const path = extractPath(ctx.args);
    if (!path) return null;
    if (!containsPathTraversal(path)) return null;
    return match(this, 'deny', `Path "${path}" contains traversal segments ("..").`, path);
  },
};

const symlinkEscape: Rule = {
  id: 'TG02-symlink-escape',
  category,
  description: 'A symlink/link operation targets a path outside the declared filesystem scope.',
  evaluate(ctx) {
    const op = extractOperation(ctx.args) ?? '';
    if (!/symlink|link/.test(op)) return null;
    const path = extractPath(ctx.args);
    if (!path) return null;
    if (isWithinScope(path, ctx.scope.filesystem)) return null;
    return match(
      this,
      'deny',
      `Symlink target "${path}" is outside the declared filesystem scope.`,
      path,
    );
  },
};

const sensitiveSystemPath: Rule = {
  id: 'TG02-sensitive-system-path',
  category,
  description: 'A write/delete targets a sensitive absolute system directory.',
  evaluate(ctx) {
    const path = extractPath(ctx.args);
    if (!path) return null;
    const op = extractOperation(ctx.args) ?? '';
    if (!WRITE_OPS.has(op) && !DELETE_OPS.has(op) && !CHMOD_OPS.has(op)) return null;
    const lower = path.toLowerCase();
    const hit = SENSITIVE_SYSTEM_PREFIXES.find((prefix) => lower.startsWith(prefix));
    if (!hit) return null;
    return match(
      this,
      'deny',
      `Target "${path}" is under a sensitive system directory (${hit}).`,
      path,
    );
  },
};

export const filesystemScopeRules: readonly Rule[] = [
  writeOutsideScope,
  deleteOutsideScope,
  chmodOutsideScope,
  readOutsideScope,
  pathTraversal,
  symlinkEscape,
  sensitiveSystemPath,
];

File: python/src/toolgovern/classifier/filesystem_scope.py
"""TG02 -- Filesystem Scope Escalation.

Ported from ``packages/toolgovern/src/classifier/filesystem-scope.ts``.

Fires when a call attempts a write, delete, or permission change outside the caller's declared
filesystem scope (``scope.filesystem``, a list of allowed path prefixes), or targets a small set
of sensitive absolute system directories regardless of scope.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, List, Optional, Sequence

from ..types import RuleContext, RuleMatch
from .util import (
    contains_path_traversal,
    extract_operation,
    extract_path,
    is_credential_granted,
    is_path_within,
    normalize_for_match,
)

_CATEGORY = "TG02"

_WRITE_OPS = {"write", "create", "append", "put", "save"}
_DELETE_OPS = {"delete", "remove", "unlink", "rm", "rmdir"}
_CHMOD_OPS = {"chmod", "chown", "setpermissions", "set_permissions"}
_READ_OPS = {"read", "get", "load", "fetch", "cat", "open"}
_SENSITIVE_SYSTEM_PREFIXES = ["/etc", "/usr", "/bin", "/sbin", "/system", "/private/etc"]


def _extract_normalized_path(args) -> Optional[str]:
    """Extracts a call's path argument and runs it through the same obfuscation-normalization
    pass TG01/TG03/TG04 already apply, so a path cannot dodge a scope or sensitive-prefix check
    by smuggling in a formatting character a literal/prefix comparison wouldn't see."""
    raw = extract_path(args)
    return normalize_for_match(raw) if raw else None


def _is_within_scope(path: str, filesystem: Sequence[str]) -> bool:
    if len(filesystem) == 0:
        return False
    return any(is_path_within(path, prefix) for prefix in filesystem)


@dataclass
class _Rule:
    id: str
    category: str
    description: str
    _evaluate: Callable[[RuleContext], Optional[RuleMatch]]

    def evaluate(self, ctx: RuleContext) -> Optional[RuleMatch]:
        return self._evaluate(ctx)


def _match(rule_id: str, decision: str, reason: str, matched_argument: str) -> RuleMatch:
    return RuleMatch(
        rule_id=rule_id,
        category=_CATEGORY,  # type: ignore[arg-type]
        decision=decision,  # type: ignore[arg-type]
        reason=reason,
        matched_argument=matched_argument,
    )


def _write_outside_scope_evaluate(ctx: RuleContext) -> Optional[RuleMatch]:
    path = _extract_normalized_path(ctx.args)
    if not path:
        return None
    op = extract_operation(ctx.args) or ("write" if "write" in ctx.tool.lower() else "")
    if op not in _WRITE_OPS:
        return None
    if _is_within_scope(path, ctx.scope.filesystem):
        return None
    return _match(
        "TG02-write-outside-scope",
        "require-approval",
        f'Write target "{path}" is outside the declared filesystem scope.',
        path,
    )


def _delete_outside_scope_evaluate(ctx: RuleContext) -> Optional[RuleMatch]:
    path = _extract_normalized_path(ctx.args)
    if not path:
        return None
    op = extract_operation(ctx.args) or ("delete" if "delete" in ctx.tool.lower() else "")
    if op not in _DELETE_OPS:
        return None
    if _is_within_scope(path, ctx.scope.filesystem):
        return None
    return _match(
        "TG02-delete-outside-scope",
        "deny",
        f'Delete target "{path}" is outside the declared filesystem scope.',
        path,
    )


def _chmod_outside_scope_evaluate(ctx: RuleContext) -> Optional[RuleMatch]:
    path = _extract_normalized_path(ctx.args)
    if not path:
        return None
    op = extract_operation(ctx.args)
    if not op or op not in _CHMOD_OPS:
        return None
    if _is_within_scope(path, ctx.scope.filesystem):
        return None
    return _match(
        "TG02-chmod-outside-scope",
        "deny",
        f'Permission change on "{path}" is outside the declared filesystem scope.',
        path,
    )


def _read_outside_scope_evaluate(ctx: RuleContext) -> Optional[RuleMatch]:
    path = _extract_normalized_path(ctx.args)
    if not path:
        return None
    op = extract_operation(ctx.args) or ("read" if "read" in ctx.tool.lower() else "")
    if op not in _READ_OPS:
        return None
    if _is_within_scope(path, ctx.scope.filesystem):
        return None
    if is_credential_granted(path, ctx.scope.credentials):
        return None
    return _match(
        "TG02-read-outside-scope",
        "require-approval",
        f'Read target "{path}" is outside the declared filesystem scope.',
        path,
    )


def _path_traversal_evaluate(ctx: RuleContext) -> Optional[RuleMatch]:
    path = _extract_normalized_path(ctx.args)
    if not path:
        return None
    if not contains_path_traversal(path):
        return None
    return _match(
        "TG02-path-traversal", "deny", f'Path "{path}" contains traversal segments ("..").', path
    )


def _symlink_escape_evaluate(ctx: RuleContext) -> Optional[RuleMatch]:
    op = extract_operation(ctx.args) or ""
    if "symlink" not in op and "link" not in op:
        return None
    path = _extract_normalized_path(ctx.args)
    if not path:
        return None
    if _is_within_scope(path, ctx.scope.filesystem):
        return None
    return _match(
        "TG02-symlink-escape",
        "deny",
        f'Symlink target "{path}" is outside the declared filesystem scope.',
        path,
    )


def _sensitive_system_path_evaluate(ctx: RuleContext) -> Optional[RuleMatch]:
    path = _extract_normalized_path(ctx.args)
    if not path:
        return None
    op = extract_operation(ctx.args) or ""
    if op not in _WRITE_OPS and op not in _DELETE_OPS and op not in _CHMOD_OPS:
        return None
    # Case-insensitive, same as the original raw-prefix check this replaces -- every entry in
    # SENSITIVE_SYSTEM_PREFIXES is already lowercase.
    lower = path.lower()
    hit = next((p for p in _SENSITIVE_SYSTEM_PREFIXES if is_path_within(lower, p)), None)
    if not hit:
        return None
    return _match(
        "TG02-sensitive-system-path",
        "deny",
        f'Target "{path}" is under a sensitive system directory ({hit}).',
        path,
    )


filesystem_scope_rules: List[_Rule] = [
    _Rule("TG02-write-outside-scope", _CATEGORY, "A write/create targets a path outside the declared filesystem scope.", _write_outside_scope_evaluate),
    _Rule("TG02-delete-outside-scope", _CATEGORY, "A delete targets a path outside the declared filesystem scope.", _delete_outside_scope_evaluate),
    _Rule("TG02-chmod-outside-scope", _CATEGORY, "A permission change targets a path outside the declared filesystem scope.", _chmod_outside_scope_evaluate),
    _Rule(
        "TG02-read-outside-scope",
        _CATEGORY,
        "A read targets a path outside the caller's declared filesystem scope. An explicitly "
        "granted credential path (scope.credentials) is not flagged here.",
        _read_outside_scope_evaluate,
    ),
    _Rule("TG02-path-traversal", _CATEGORY, 'A path uses ".." segments that could escape a scoped prefix.', _path_traversal_evaluate),
    _Rule("TG02-symlink-escape", _CATEGORY, "A symlink/link operation targets a path outside the declared filesystem scope.", _symlink_escape_evaluate),
    _Rule("TG02-sensitive-system-path", _CATEGORY, "A write/delete targets a sensitive absolute system directory.", _sensitive_system_path_evaluate),
]
